snuba/clickhouse/http.py

```diff
@@ -35,75 +40,60 @@ def encode(self, value: WriterTableRow) -> JSONRow:
+        pool: HTTPConnectionPool,
+        database: str,
         table_name: str,
-        host: str,
-        port: int,
         user: str,
         password: str,
-        database: str,
-        metrics: MetricsBackend,
-        options: Optional[Mapping[str, Any]] = None,
-        chunk_size: Optional[int] = 1,
-    ):
-        """
-        Builds a writer to send a batch to Clickhouse.
-        The encoder function will be applied to each row to turn it into bytes.
-        We send data to the server with Transfer-Encoding: chunked. If chunk size is 0
-        we send the entire content in one chunk, otherwise it is the rows per chunk.
-        """
-        self.__pool = HTTPConnectionPool(host, port)
-        self.__options = options if options is not None else {}
-        self.__table_name = table_name
-        self.__chunk_size = chunk_size
-        self.__user = user
-        self.__password = password
-        self.__database = database
-        self.__metrics = MetricsWrapper(metrics, "writer", {"table_name": table_name})
-
-    def _prepare_chunks(self, rows: Iterable[JSONRow]) -> Iterable[bytes]:
-        total_bytes_size = 0
-        chunk = []
-
-        for row in rows:
-            chunk.append(row)
-            if self.__chunk_size and len(chunk) == self.__chunk_size:
-                chunk_bytes = b"".join(chunk)
-                yield chunk_bytes
-                self.__metrics.timing("chunk.size", len(chunk_bytes))
-                total_bytes_size += len(chunk_bytes)
-                chunk = []
-
-        if chunk:
-            chunk_bytes = b"".join(chunk)
-            yield chunk_bytes
-            self.__metrics.timing("chunk.size", len(chunk_bytes))
-            total_bytes_size += len(chunk_bytes)
+        options: Mapping[str, Any],  # should be ``Mapping[str, str]``?
+    ) -> None:
+        read_fd, write_fd = os.pipe()
+        self.__input = open(write_fd, "wb")
 
-        self.__metrics.timing("total.size", total_bytes_size)
-
-    def write(self, values: Iterable[JSONRow]) -> None:
-        response = self.__pool.urlopen(
+        self.__result = executor.submit(
+            pool.urlopen,
             "POST",
             "/?"
             + urlencode(
                 {
-                    **self.__options,
-                    "query": f"INSERT INTO {self.__database}.{self.__table_name} FORMAT JSONEachRow",
+                    **options,
+                    "query": f"INSERT INTO {database}.{table_name} FORMAT JSONEachRow",
                 }
             ),
             headers={
-                "X-ClickHouse-User": self.__user,
-                "X-ClickHouse-Key": self.__password,
+                "X-ClickHouse-User": user,
+                "X-ClickHouse-Key": password,
                 "Connection": "keep-alive",
                 "Accept-Encoding": "gzip,deflate",
             },
-            body=self._prepare_chunks(values),
-            chunked=True,
+            body=open(read_fd, "rb"),
```

Comment:
Isn't the approach of using a file like body here adding an additional copy of each JSONRow we send to clickhouse than what we need (I guess you are still having the same number of copies we have today)?
I am not 100% certain on the `urlopen` would do in this case, but the pipe itself should be doing a copy of every record.
Wouldn't we be avoiding a copy by using a synchronized queue to send rows to the thread that runs `urlopen` and building a file like object that reads from the synchronized queue, and passing this object to urlopen? 
Specifically:
- implement a file like object, this object reads records from a Queue passed in the constructor.
- The append method of this class just writes the JSONRows on the queue
- We pass the file object above to urlopen.

Reply:
I assumed (but did not test) that dealing with the OS pipe buffer would be more efficient than acquiring the queue lock on every write and read. I think it'd be worth testing both of these approaches though.

Reply:
I compared the two approaches by writing a batch of ~65k values that were ~8kb each ([script](https://gist.github.com/tkaemming/c5b4133110222fc68c08ec4f35579b5a)), and they showed roughly equivalent performance on my laptop (within about ~5% of each other, the threading one seems like it might be marginally faster.) I'll update this to use the queue-backed implementation instead as it's a bit easier to read/understand and doesn't depend on outside factors like the pipe buffer size.